### ISSUE TITLE
PersistentFSMActor and non persistent events

### ISSUE THREAD

**Issue by sirocchj** [akka/akka#18105]
While using `PersistentFsmActor` I discovered that I cannot model my persistent actor state's data as a single class that gets updated by *both* persistent events *and* non-persistent ones (i.e. one container to cater for both persistent and transient state) as state data can only mutate when applying domain events to it, hence I was forced to keep the transient state in vars inside my actor.

This may be unsound but I'd like to see if `using` (or a variation of `applying`) could accomplish what I'd be aiming for.

Any feedback would be much appreciated

Discussion started on https://groups.google.com/d/msg/akka-user/DNG5SZcPffw/KI4x4IFZDAAJ

**Comment by leonidb** [comment]
Can be implemented by maintaining both non-persistent Data and DomainData separately. 
User will be able to set the non-persistent Data with ```using```, as in the original ```FSM``` actor DSL.
For example:
```Scala
goto(Shopping) using(non-persistent data) applying ItemAdded(item)
```
It will add additional complexity both to API and usage, as the non-persistent Data will be closely related to the State, while DomainData will be derived from Domain Events and sort of "decoupled" from current State.

**Comment by patriknw** [comment]
I'm not convinced that it is worth the additional complexity. The alternative is to keep the non-persistent data in a separate `var` in the actor. Please convince me if you disagree.

**Comment by sirocchj** [comment]
Thanks @patriknw and @leonidb for the feedback

I've implemented my solution, as in the ticket description, using `var`s in my persistent actor so I'm totally fine with it. However, it feels a bit clumsy, having to maintain state data in a completely separate instance, even though it is tightly related to the persistent state. Hence, this ticket.

I've been playing w/ it a bit (adding another `applying` which takes transient events and does not attempt to persist them on state transitions) but it feels, as you both rightly say, like I've added complexity for little gain.
Furthermore, it happens in my case that the actual *state* is the same (i.e. I want to `stay()` while adding some transient state data, related to the persistent one), but you may well imagine that that is not always necessarily true (others may then want to `goto(someNonPersistentState)`).

Bottom line, I am starting to believe that:
1. I should consider having another actor, `with FSM`, that deals with transient state and reports back to persistent one once "done" (in case I need more than transient state data)
2. if you see fit for the co-existence of persistent and transient state and state data within a *single* actor, there should be a way of mixing in both PersistentFSM and FSM in some neat way (I'm thinking about state monad + lenses)

**Comment by ktoso** [comment]
We thought this one through and it seems it would complicate the API too much (needs additional param for the "non persistent state" and additional "using" methods which would be confusing again).

Seems we're in agreement there's different ways of implementing this without a very complex API (child actor perhaps), so closing as will-not-fix.
